### PR TITLE
Issue 4 - As a user I would like to be able to log into the private section using my Google account or an email

### PR DESCRIPTION
**Pull request by jaysonbucy**
Going to open a bug to include gravatar for user image and to create a method to add a link to an image for the user image.

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -1,46 +1,67 @@
...
+
 hr.shadow{
   height: 10px;
   border: 0;
-  box-shadow: 0 10px 10px -10px #8c8b8b inset;
+  box-shadow: 0 10px 10px -10px $dark-green inset;
+}
+
+.view-container{
+  padding-left: 2em;
```
Comment (kylessnell) [inline app/assets/stylesheets/application.scss]: personally prefer using `em`s for font related spacing, and `px`s or `%`s for positioning

File: app/helpers/user_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module UserHelper
```
Comment (kylessnell) [inline app/helpers/user_helper.rb]: need this?

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.firstname = auth.info.first_name
+      user.lastname = auth.info.last_name
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0,20]
+      user.image = auth.info.image
+      # If you are using confirmable and the provider(s) you use validate emails,
+      # uncomment the line below to skip the confirmation emails.
+      # user.skip_confirmation!
```
Comment (kylessnell) [inline app/models/user.rb]: need these comments?

File: app/views/devise/registrations/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    <br>
+    <button type='button' class="btn btn-primary">
+      <%= f.submit "Update" %>
+    </button>
+  <% end %>
+
+  <!-- <h3>Cancel my account</h3>
+
+  <p>Unhappy? <%= button_to "Cancel my account", registration_path(resource_name), data: { confirm: "Are you sure?" }, method: :delete %></p> -->
```
Comment (kylessnell) [inline app/views/devise/registrations/edit.html.erb:53]: need this?

File: config/application.config.yml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+GOOGLE_APP_ID:
+GOOGLE_APP_SECRET:
```
Comment (kylessnell) [inline config/application.config.yml:2]: is this meant to be gitignored?

File: config/initializers/devise.rb [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+  # confirmation, reset password and unlock tokens in the database.
+  # Devise will use the `secret_key_base` as its `secret_key`
+  # by default. You can change it below and use your own secret key.
+  # config.secret_key = 'bbb689ac4c6683f17ffda6234e3bfeee3b77b7f0ae4834c41be79a38ebd19a8d96b3efa9ccc358b92c65a9300099c8c054dc4f05e46ebaa219c06ec0bc63b46d'
+
+  # ==> Mailer Configuration
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Comment (kylessnell) [inline config/initializers/devise.rb:15]: might need to update this :)

File: spec/helpers/user_helper_spec.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
Comment (kylessnell) [inline spec/helpers/user_helper_spec.rb]: need this?

File: spec/views/user/index.html.erb_spec.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "user/index.html.erb", type: :view do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (kylessnell) [inline spec/views/user/index.html.erb_spec.rb]: where are you tests for this view?

File: spec/views/user/show.html.erb_spec.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "user/show.html.erb", type: :view do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (kylessnell) [inline spec/views/user/show.html.erb_spec.rb]: tests?

File: app/helpers/user_helper.rb (hunk 2 above)
Comment (jaysonbucy) [reply app/helpers/user_helper.rb]: nope, deleting

File: app/models/user.rb (hunk 3 above)
Comment (jaysonbucy) [reply app/models/user.rb]: not yet, I've got to talk to the customer and see if they want the user to get email confirmation

File: app/views/devise/registrations/edit.html.erb (hunk 4 above)
Comment (jaysonbucy) [reply app/views/devise/registrations/edit.html.erb:53]: Going to see if the customer wants to give the user the option

File: config/application.config.yml (hunk 5 above)
Comment (jaysonbucy) [reply config/application.config.yml:2]: no, that is so other developers know what they need in the actual `application.yml` file without housing the actual values.

File: config/initializers/devise.rb (hunk 6 above)
Comment (jaysonbucy) [reply config/initializers/devise.rb:15]: Auto-generated, I'll look at it

File: spec/helpers/user_helper_spec.rb (hunk 7 above)
Comment (jaysonbucy) [reply spec/helpers/user_helper_spec.rb]: nope deleting

File: spec/views/user/index.html.erb_spec.rb (hunk 8 above)
Comment (jaysonbucy) [reply spec/views/user/index.html.erb_spec.rb]: historically I've not written view tests

File: spec/views/user/show.html.erb_spec.rb (hunk 9 above)
Comment (jaysonbucy) [reply spec/views/user/show.html.erb_spec.rb]: same as above

File: app/assets/stylesheets/application.scss [hunk 10]
```diff
@@ -1,46 +1,67 @@
 @import "bootstrap-sprockets";
 @import "bootstrap";
```
Comment (kylessnell) [inline app/assets/stylesheets/application.scss:3]: having all of your CSS in one gigantic file is going to become really hard to deal with.

Comment (jaysonbucy) [reply app/assets/stylesheets/application.scss:3]: I'm hoping there won't be much more than this, if it gets much larger than this I will break it up. But right now, I can't really see a smooth breaking point.

File: app/views/welcome/_control_bar.html.erb [hunk 11]
```diff
@@ -1,23 +1,23 @@
 <div class="header-control-bar">
   <% if current_page?(root_url) %>
```
Comment (kylessnell) [inline app/views/welcome/_control_bar.html.erb:2]: you don't have tests for any of these `if` statements about which buttons to display

File: app/views/users/_control_bar.html.erb [hunk 12]
```diff
@@ -0,0 +1,16 @@
+<div class="header-control-bar">
+  <% if current_page?(root_url) %>
```
Comment (kylessnell) [inline app/views/users/_control_bar.html.erb:2]: no tests for this logic

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -27,7 +38,11 @@
         <h1>Greater Ada Swim Club</h1>
       </div>
     </div>
-    <%= render "control_bar" %>
+    <% if user_signed_in? %>
```
Comment (kylessnell) [inline app/views/layouts/application.html.erb:41]: no tests for this logic

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -18,7 +18,18 @@
           <%= image_tag('gasc_header_2015.png', width: 200, alt: 'GASC') %>
         </div>
         <div class="col-md-3 col-sm-4 sign-in">
-          Sign In
+          <% if user_signed_in? %>
```
Comment (kylessnell) [inline app/views/layouts/application.html.erb:21]: no tests for this stuff

File: app/views/devise/registrations/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+    <div class="field">
+      <%= f.label :email %><br />
+      <%= f.email_field :email, autofocus: true %>
+    </div>
+
+    <div class="field">
+      <%= f.label :password %>
+      <% if @minimum_password_length %>
+      <em>(<%= @minimum_password_length %> characters minimum)</em>
```
Comment (kylessnell) [inline app/views/devise/registrations/new.html.erb:25]: tests for the error states for this form?

File: app/views/devise/registrations/edit.html.erb [hunk 16]
```diff
@@ -0,0 +1,56 @@
+<div class="view-container">
+  <h2>Edit <%= resource_name.to_s.humanize %></h2>
+
+  <%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f| %>
+    <%= devise_error_messages! %>
+
+    <div class="field">
```
Comment (kylessnell) [inline app/views/devise/registrations/edit.html.erb:7]: registration flow tests?

File: app/views/devise/passwords/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :post }) do |f| %>
+  <%= devise_error_messages! %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Send me reset password instructions" %>
```
Comment (kylessnell) [inline app/views/devise/passwords/new.html.erb:12]: test for this?

**Pull request merged**